File: backend/app/schemas/analytics.py
"""Analytics API schemas."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

LotteryType = Literal["ssq", "dlt"]


class AnalyticsQuery(BaseModel):
    lottery_type: LotteryType = "ssq"
    window: int | None = Field(default=50, ge=1, le=5000)


class NumberStat(BaseModel):
    number: int
    count: int
    ratio: float | None = None
    missing: int | None = None
    last_issue: str | None = None


class AnalyticsOverviewData(BaseModel):
    lottery_type: LotteryType
    metrics: dict[str, Any]
    hot_cold: dict[str, Any]
    frequency_primary: list[dict[str, Any]]
    missing_primary: list[dict[str, Any]]
    sum_span: list[dict[str, Any]]
    zones: list[dict[str, Any]]
    cooccurrence: list[dict[str, Any]]

File: frontend/src/types/analytics.ts
export type AnalyticsOverview = {
  lottery_type: "ssq" | "dlt";
  metrics: {
    total_draws: number;
    latest_issue: string | null;
    latest_draw_date: string | null;
    avg_sum: number | null;
    avg_span: number | null;
  };
  hot_cold: {
    window: number;
    hot: Array<{ number: number; count: number; ratio: number }>;
    cold: Array<{ number: number; count: number; ratio: number }>;
  };
  frequency_primary: Array<{ number: number; count: number; ratio: number }>;
  missing_primary: Array<{ number: number; missing: number; last_issue: string | null }>;
  sum_span: Array<{
    issue: string;
    draw_date: string;
    sum: number;
    span: number;
    odd: number;
    even: number;
    odd_even: string;
  }>;
  zones: Array<{
    issue: string;
    draw_date: string;
    zone_low: number;
    zone_mid: number;
    zone_high: number;
    pattern: string;
  }>;
  cooccurrence: Array<{ a: number; b: number; count: number }>;
};
